<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 9 Practical: Principal component analysis | Essentials of Mathematics and Statistics</title>
<meta name="author" content="Anas A Rana">
<meta name="description" content="In this practical we will practice some of the ideas outlined in the lecture on Principal Component Analysis (PCA), this will include computing principal components, visualisation techniques and...">
<meta name="generator" content="bookdown 0.40 with bs4_book()">
<meta property="og:title" content="Chapter 9 Practical: Principal component analysis | Essentials of Mathematics and Statistics">
<meta property="og:type" content="book">
<meta property="og:description" content="In this practical we will practice some of the ideas outlined in the lecture on Principal Component Analysis (PCA), this will include computing principal components, visualisation techniques and...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 9 Practical: Principal component analysis | Essentials of Mathematics and Statistics">
<meta name="twitter:description" content="In this practical we will practice some of the ideas outlined in the lecture on Principal Component Analysis (PCA), this will include computing principal components, visualisation techniques and...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><link href="libs/Manrope-0.4.9/font.css" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Fira%20Code:wght@100&amp;display=swap" rel="stylesheet">
<link href="libs/Spectral-0.4.9/font.css" rel="stylesheet">
<script src="libs/bs3compat-0.8.0/transition.js"></script><script src="libs/bs3compat-0.8.0/tabs.js"></script><script src="libs/bs3compat-0.8.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Essentials of Mathematics and Statistics</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html">Introduction to the Practical</a></li>
<li class="book-part">Probability in R</li>
<li><a class="" href="simulating-random-numbers.html"><span class="header-section-number">1</span> Simulating random numbers</a></li>
<li><a class="" href="markov-chains.html"><span class="header-section-number">2</span> Markov Chains</a></li>
<li><a class="" href="a-monopoly-simulation.html"><span class="header-section-number">3</span> A Monopoly simulation</a></li>
<li><a class="" href="monte-carlo-methods.html"><span class="header-section-number">4</span> Monte Carlo Methods</a></li>
<li><a class="" href="maximum-likelihood.html"><span class="header-section-number">5</span> Maximum Likelihood</a></li>
<li><a class="" href="sol-mle.html">Solution: Model Answers: MLE</a></li>
<li><a class="" href="confidence-intervals.html"><span class="header-section-number">6</span> Confidence Intervals</a></li>
<li><a class="" href="sol-ex-CI.html">Model Answer: Confidence Interval</a></li>
<li><a class="" href="computational-testing-techniques.html"><span class="header-section-number">7</span> Computational Testing Techniques</a></li>
<li class="book-part">Statistical Modelling</li>
<li><a class="" href="practical-linear-regression.html"><span class="header-section-number">8</span> Practical: Linear regression</a></li>
<li><a class="active" href="practical-principal-component-analysis.html"><span class="header-section-number">9</span> Practical: Principal component analysis</a></li>
<li><a class="" href="practical-multiple-regression.html"><span class="header-section-number">10</span> Practical: Multiple regression</a></li>
<li><a class="" href="practical-generalised-linear-models.html"><span class="header-section-number">11</span> Practical: Generalised linear models</a></li>
<li class="book-part">Solutions</li>
<li><a class="" href="simSol.html"><span class="header-section-number">A</span> Solution: Simulations</a></li>
<li><a class="" href="mc-solution.html"><span class="header-section-number">B</span> Solutions: 3 state Markov Chain</a></li>
<li><a class="" href="monopolySol.html"><span class="header-section-number">C</span> Solutions: Monopoly</a></li>
<li><a class="" href="solution-monte-carlo.html">Solution: Monte Carlo</a></li>
<li><a class="" href="model-answers-computational-testing.html"><span class="header-section-number">D</span> Model Answers: Computational Testing</a></li>
<li><a class="" href="solutions-linear-regression-and-correlation.html"><span class="header-section-number">E</span> Solutions: Linear regression and Correlation</a></li>
<li class="book-part">APPENDIX</li>
<li><a class="" href="start.html">Getting started in R and Rstudio</a></li>
<li><a class="" href="data-sets.html">Data sets</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/anasrana/ems-practicals">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="practical-principal-component-analysis" class="section level1" number="9">
<h1>
<span class="header-section-number">9</span> Practical: Principal component analysis<a class="anchor" aria-label="anchor" href="#practical-principal-component-analysis"><i class="fas fa-link"></i></a>
</h1>
<p>In this practical we will practice some of the ideas outlined in the lecture on Principal Component Analysis (PCA), this will include computing principal components, visualisation techniques and an application to real data.</p>
<div id="data-1" class="section level2" number="9.1">
<h2>
<span class="header-section-number">9.1</span> Data<a class="anchor" aria-label="anchor" href="#data-1"><i class="fas fa-link"></i></a>
</h2>
<p>For this practical you will require three datasets, more information about downloading them <a href="data-sets.html#data-sets">here</a>. If you want to download them manually you can use the following links:</p>
<ul>
<li>
<code>Pollen2014.txt</code> (<a href="https://github.com/anasrana/ems-practicals/raw/main/data/Pollen2014.txt">download</a>)</li>
<li>
<code>SupplementaryLabels.txt</code> (<a href="https://github.com/anasrana/ems-practicals/raw/main/data/SupplementaryLabels.txt">download</a>)</li>
</ul>
</div>
<div id="introduction" class="section level2" number="9.2">
<h2>
<span class="header-section-number">9.2</span> Introduction<a class="anchor" aria-label="anchor" href="#introduction"><i class="fas fa-link"></i></a>
</h2>
<p>We use PCA in order to explore complex datasets. By performing dimensionality reduction we can better visualize the data that has many variables. This technique is probably the most popular tool applied across bioscience problems (e.g. for gene expression problems).</p>
<p>In many real-world dataset we deal with a high dimensional data, e.g. for a number of individuals we can take a number of health related measurement (called variables). This is great, however having a large number of variables also means that it is difficult to plot the data as it is (in its “<em>raw</em>” format), and in turn it might be difficult to understand if this dataset contains any interesting patterns/trends/relationships across individuals. Using PCA we visualize such data in a more “<em>human friendly</em>” fashion.</p>
<p>Recall:</p>
<ul>
<li>PCA performs a linear transformation to data.</li>
<li>This means that any input data can be visualized in a new coordinate system. The first coordinate (PC 1) variance is found on the first coordinate; each subsequent coordinate is orthogonal to the previous one and contains the larges variance from what was left.</li>
<li>Each principal component is associated with certain percentage of the total variation in the dataset.</li>
<li>If variables are strongly correlated with one another, a first few principal components will enable us to visualize the relationships present in any dataset.</li>
<li>Eigenvectors describe new directions, whereas accompanying eigenvalues tell us how much variance there is in the data in given direction.</li>
<li>The eigenvector with the highest eigenvalue is called the first principal component. The second highest eigenvalue would correspond to a second principle component and etc.</li>
<li>There exist a <span class="math inline">\(d\)</span> number of eigenvalues and eigenvectors; <span class="math inline">\(d\)</span> is also equal to the size of the data (number of dimensions).</li>
<li>For the purpose of visualization we preselect the first <span class="math inline">\(q\)</span> components, where <span class="math inline">\(q &lt; d\)</span>.</li>
</ul>
</div>
<div id="example-pca-on-the-mtcars-dataset" class="section level2" number="9.3">
<h2>
<span class="header-section-number">9.3</span> Example: PCA on the <code>mtcars</code> dataset<a class="anchor" aria-label="anchor" href="#example-pca-on-the-mtcars-dataset"><i class="fas fa-link"></i></a>
</h2>
<p>There are many datasets built into <code>R</code>. Wed will look at the <code>mtcars</code> dataset. Type <code><a href="https://rdrr.io/r/datasets/mtcars.html">?mtcars</a></code> to get a description of data. Then use <code><a href="https://rdrr.io/r/utils/head.html">head()</a></code> function to have a look at the first few rows; and <code><a href="https://rdrr.io/r/base/dim.html">dim()</a></code> to get the dimensions of the data.</p>
<div class="sourceCode" id="cb106"><pre class="sourceCode numberSource r numberLines"><code class="sourceCode r"><span id="cb106-1"><a href="practical-principal-component-analysis.html#cb106-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb106-2"><a href="practical-principal-component-analysis.html#cb106-2"></a><span class="fu">head</span>(mtcars)</span></code></pre></div>
<pre class="bg-info"><code>#&gt;                    mpg cyl disp  hp drat    wt  qsec vs am
#&gt; Mazda RX4         21.0   6  160 110 3.90 2.620 16.46  0  1
#&gt; Mazda RX4 Wag     21.0   6  160 110 3.90 2.875 17.02  0  1
#&gt; Datsun 710        22.8   4  108  93 3.85 2.320 18.61  1  1
#&gt; Hornet 4 Drive    21.4   6  258 110 3.08 3.215 19.44  1  0
#&gt; Hornet Sportabout 18.7   8  360 175 3.15 3.440 17.02  0  0
#&gt; Valiant           18.1   6  225 105 2.76 3.460 20.22  1  0
#&gt;                   gear carb
#&gt; Mazda RX4            4    4
#&gt; Mazda RX4 Wag        4    4
#&gt; Datsun 710           4    1
#&gt; Hornet 4 Drive       3    1
#&gt; Hornet Sportabout    3    2
#&gt; Valiant              3    1</code></pre>
<div class="sourceCode" id="cb108"><pre class="sourceCode numberSource r numberLines"><code class="sourceCode r"><span id="cb108-1"><a href="practical-principal-component-analysis.html#cb108-1"></a><span class="fu">dim</span>(mtcars)</span></code></pre></div>
<pre class="bg-info"><code>#&gt; [1] 32 11</code></pre>
<p>In this case we have <span class="math inline">\(32\)</span> examples (cars in this case), and <span class="math inline">\(11\)</span> features.
Now we can perform a principal component analysis, in <code>R</code> it is implemented as the <code><a href="https://rdrr.io/r/stats/prcomp.html">prcomp()</a></code> function. We can type <code><a href="https://rdrr.io/r/stats/prcomp.html">?prcomp</a></code> to see a description of the function and some help on possible arguments. Here we set <code>center</code> and <code>scale</code> arguments to <code>TRUE</code>, recall from the lecture why this is important. We can try to perform PCA without scaling and centering and compare.</p>
<div class="sourceCode" id="cb110"><pre class="sourceCode numberSource r numberLines"><code class="sourceCode r"><span id="cb110-1"><a href="practical-principal-component-analysis.html#cb110-1"></a>cars_pca <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(mtcars, <span class="at">center =</span> <span class="cn">TRUE</span>, <span class="at">scale =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>We can use the <code><a href="https://rdrr.io/r/base/summary.html">summary()</a></code> function to summarise the results from PCA, it will return the standard deviation, the proportion of variance explained by each principal component, and the cumulative proportion.</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode numberSource r numberLines"><code class="sourceCode r"><span id="cb111-1"><a href="practical-principal-component-analysis.html#cb111-1"></a>pca_summary <span class="ot">&lt;-</span> <span class="fu">summary</span>(cars_pca)</span>
<span id="cb111-2"><a href="practical-principal-component-analysis.html#cb111-2"></a><span class="fu">print</span>(pca_summary)</span></code></pre></div>
<pre class="bg-info"><code>#&gt; Importance of components:
#&gt;                           PC1    PC2     PC3     PC4
#&gt; Standard deviation     2.5707 1.6280 0.79196 0.51923
#&gt; Proportion of Variance 0.6008 0.2409 0.05702 0.02451
#&gt; Cumulative Proportion  0.6008 0.8417 0.89873 0.92324
#&gt;                            PC5     PC6    PC7     PC8
#&gt; Standard deviation     0.47271 0.46000 0.3678 0.35057
#&gt; Proportion of Variance 0.02031 0.01924 0.0123 0.01117
#&gt; Cumulative Proportion  0.94356 0.96279 0.9751 0.98626
#&gt;                           PC9    PC10   PC11
#&gt; Standard deviation     0.2776 0.22811 0.1485
#&gt; Proportion of Variance 0.0070 0.00473 0.0020
#&gt; Cumulative Proportion  0.9933 0.99800 1.0000</code></pre>
<p><em>Note,</em> <code>Proportion of Variance</code> will always add up to <span class="math inline">\(1\)</span>. Here the <code>PC1</code> explain <span class="math inline">\(60.08%\)</span> of the variance, and <code>PC2</code> explains <span class="math inline">\(24.09%\)</span>, which means together <code>PC1</code> and <code>PC2</code> account for <span class="math inline">\(84.17%\)</span> of the variance.</p>
<p>Using the <code><a href="https://rdrr.io/r/utils/str.html">str()</a></code> function we can see the full structure of an <code>R</code> object, or alternatively using <code><a href="https://rdrr.io/r/stats/prcomp.html">?prcomp</a></code> in the <em>Value</em> section. In this case the <code>cars_pca</code> variable is a list containing several variables, <code>x</code> is the data represented using the new principal components. We can now plot the data in the first two principal components:</p>
<div class="sourceCode" id="cb113"><pre class="sourceCode numberSource r numberLines"><code class="sourceCode r"><span id="cb113-1"><a href="practical-principal-component-analysis.html#cb113-1"></a>pca_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(cars_pca<span class="sc">$</span>x, <span class="at">make =</span> stringr<span class="sc">::</span><span class="fu">word</span>(<span class="fu">rownames</span>(mtcars), <span class="dv">1</span>))</span>
<span id="cb113-2"><a href="practical-principal-component-analysis.html#cb113-2"></a></span>
<span id="cb113-3"><a href="practical-principal-component-analysis.html#cb113-3"></a><span class="fu">ggplot</span>(pca_df, <span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, <span class="at">col =</span> make)) <span class="sc">+</span></span>
<span id="cb113-4"><a href="practical-principal-component-analysis.html#cb113-4"></a><span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">3</span>) <span class="sc">+</span></span>
<span id="cb113-5"><a href="practical-principal-component-analysis.html#cb113-5"></a><span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"PC1 60.08%"</span>,</span>
<span id="cb113-6"><a href="practical-principal-component-analysis.html#cb113-6"></a>     <span class="at">y =</span> <span class="st">"PC2 24.09 %"</span>,</span>
<span id="cb113-7"><a href="practical-principal-component-analysis.html#cb113-7"></a>     <span class="at">title =</span> <span class="st">"Principal components for mtcars"</span>) <span class="sc">+</span></span>
<span id="cb113-8"><a href="practical-principal-component-analysis.html#cb113-8"></a><span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span></code></pre></div>
<div class="inline-figure"><img src="PCA_files/figure-html/plot_2d-1.png" width="95%" style="display: block; margin: auto;"></div>
<p>Here we added a color based on the make of each car. We can observe which samples (or cars) cluster together. Have a look at these variables and decide why certain cars or models would cluster together.</p>
<p>We created this plot using the <code>ggplot2</code> package, it is also possible to do this using base plot if you prefer.</p>
<div class="sourceCode" id="cb114"><pre class="sourceCode numberSource r numberLines"><code class="sourceCode r"><span id="cb114-1"><a href="practical-principal-component-analysis.html#cb114-1"></a><span class="fu">plot</span>(pca_df<span class="sc">$</span>PC1, pca_df<span class="sc">$</span>PC2)</span></code></pre></div>
</div>
<div id="creating-plots-for-pca" class="section level2" number="9.4">
<h2>
<span class="header-section-number">9.4</span> Creating plots for PCA<a class="anchor" aria-label="anchor" href="#creating-plots-for-pca"><i class="fas fa-link"></i></a>
</h2>
<p>Next we look at another representation of the data, the <em>biplot</em>. This is a combination of a PCA plot of the data and a <em>score plot</em>. We saw the PCA plot in the previous section in a <em>biplot</em> we add the original axis as arrows.</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode numberSource r numberLines"><code class="sourceCode r"><span id="cb115-1"><a href="practical-principal-component-analysis.html#cb115-1"></a><span class="fu">biplot</span>(cars_pca)</span></code></pre></div>
<div class="inline-figure"><img src="PCA_files/figure-html/biplot-1.png" width="95%" style="display: block; margin: auto;"></div>
<p>We can see the original axis starting from the origin. Therefore we can make observations about the original variables (e.g. <code>cyl</code> and <code>mpg</code> contribute to PC1) and how the data points relates to these axes.</p>
</div>
<div id="exercise-i" class="section level2 infobox" number="9.5">
<h2>
<span class="header-section-number">9.5</span> Exercise I<a class="anchor" aria-label="anchor" href="#exercise-i"><i class="fas fa-link"></i></a>
</h2>
<p>Now try to perform a PCA on the <code>USArrests</code> data also build into <code>R</code>. Typing <code><a href="https://rdrr.io/r/datasets/USArrests.html">?USArrests</a></code> will give you further information on the data. Perform the analysis on the subset <code>USArrests[, -3]</code> data.</p>
</div>
<div id="example-single-cell-data" class="section level2" number="9.6">
<h2>
<span class="header-section-number">9.6</span> Example: Single cell data<a class="anchor" aria-label="anchor" href="#example-single-cell-data"><i class="fas fa-link"></i></a>
</h2>
<p>We can now try to apply what we learned above on a more realistic datasets. You can download the data either on <em>canvas</em> or using these links <a href="https://raw.github.com/anasrana/module1-practical_Bham/master/data/Pollen2014.txt"><code>Pollen2014.txt</code></a> and <a href="https://raw.github.com/anasrana/module1-practical_Bham/master/data/SupplementaryLabels.txt"><code>SupplementaryLabels.txt</code></a>. Her we will be dealing with single cell RNA-Seq (scRNA-Seq) data, which consist of <span class="math inline">\(300\)</span> single cells measured across <span class="math inline">\(8686\)</span> genes.</p>
<div class="sourceCode" id="cb116"><pre class="sourceCode numberSource r numberLines"><code class="sourceCode r"><span id="cb116-1"><a href="practical-principal-component-analysis.html#cb116-1"></a>pollen_df <span class="ot">&lt;-</span><span class="fu">read.table</span>(<span class="st">"Pollen2014.txt"</span>, <span class="at">sep=</span><span class="st">','</span>, <span class="at">header =</span> T,<span class="at">row.names=</span><span class="dv">1</span>)</span>
<span id="cb116-2"><a href="practical-principal-component-analysis.html#cb116-2"></a></span>
<span id="cb116-3"><a href="practical-principal-component-analysis.html#cb116-3"></a>label_df <span class="ot">&lt;-</span><span class="fu">read.table</span>(<span class="st">"SupplementaryLabels.txt"</span>, <span class="at">sep=</span><span class="st">','</span>, <span class="at">header =</span> T)</span>
<span id="cb116-4"><a href="practical-principal-component-analysis.html#cb116-4"></a></span>
<span id="cb116-5"><a href="practical-principal-component-analysis.html#cb116-5"></a>pollen_df[<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>]</span></code></pre></div>
<pre class="bg-info"><code>#&gt;          Cell_2338_1 Cell_2338_10 Cell_2338_11 Cell_2338_12
#&gt; MTND2P28          78          559          811          705
#&gt; MTATP6P1        2053         1958         4922         4409
#&gt; NOC2L              1          125          126            0
#&gt; ISG15           2953         4938          580          523
#&gt; CPSF3L             2           42           19            0
#&gt; MXRA8              0            0            0            0
#&gt; AURKAIP1         302          132           64          492
#&gt; CCNL2              0          235            0           84
#&gt; MRPL20           330          477          288          222
#&gt; SSU72            604          869         2046          158
#&gt;          Cell_2338_13 Cell_2338_14
#&gt; MTND2P28          384          447
#&gt; MTATP6P1         2610         3709
#&gt; NOC2L             487           66
#&gt; ISG15            2609            1
#&gt; CPSF3L             37           12
#&gt; MXRA8               0            0
#&gt; AURKAIP1           11          182
#&gt; CCNL2              13           11
#&gt; MRPL20             44          282
#&gt; SSU72             530          272</code></pre>
<div class="sourceCode" id="cb118"><pre class="sourceCode numberSource r numberLines"><code class="sourceCode r"><span id="cb118-1"><a href="practical-principal-component-analysis.html#cb118-1"></a><span class="fu">dim</span>(pollen_df)</span></code></pre></div>
<pre class="bg-info"><code>#&gt; [1] 8686  300</code></pre>
<p>Measurements of scRNA-Seq data are integer counts, this data does not have good properties so we perform a transformation on the data. The most commonly used transformation on RNA-Seq count data is <span class="math inline">\(\log_2\)</span>. We will also transpose the data matrix to rows representing cells and columns representing genes. This is the data we can use to perform PCA.</p>
<div class="sourceCode" id="cb120"><pre class="sourceCode numberSource r numberLines"><code class="sourceCode r"><span id="cb120-1"><a href="practical-principal-component-analysis.html#cb120-1"></a><span class="co"># scRNA-Seq data transformation</span></span>
<span id="cb120-2"><a href="practical-principal-component-analysis.html#cb120-2"></a>pollen_mat <span class="ot">&lt;-</span> <span class="fu">log2</span>(<span class="fu">as.matrix</span>(pollen_df) <span class="sc">+</span> <span class="dv">1</span>)</span>
<span id="cb120-3"><a href="practical-principal-component-analysis.html#cb120-3"></a><span class="co"># transpose the data</span></span>
<span id="cb120-4"><a href="practical-principal-component-analysis.html#cb120-4"></a>pollen_mat <span class="ot">&lt;-</span> <span class="fu">t</span>(pollen_mat)</span></code></pre></div>
<p>We will now use information that we read into the <code>label_df</code> variable to rename cells.</p>
<div class="sourceCode" id="cb121"><pre class="sourceCode numberSource r numberLines"><code class="sourceCode r"><span id="cb121-1"><a href="practical-principal-component-analysis.html#cb121-1"></a><span class="co"># Check which columns we have available</span></span>
<span id="cb121-2"><a href="practical-principal-component-analysis.html#cb121-2"></a><span class="fu">colnames</span>(label_df)</span></code></pre></div>
<pre class="bg-info"><code>#&gt; [1] "Cell_Identifier"        "Population"            
#&gt; [3] "Cell_names"             "TrueLabel_CellLevel"   
#&gt; [5] "Tissue_name"            "TrueLabel_TissuelLevel"</code></pre>
<div class="sourceCode" id="cb123"><pre class="sourceCode numberSource r numberLines"><code class="sourceCode r"><span id="cb123-1"><a href="practical-principal-component-analysis.html#cb123-1"></a><span class="co"># rename rows</span></span>
<span id="cb123-2"><a href="practical-principal-component-analysis.html#cb123-2"></a><span class="fu">rownames</span>(pollen_mat) <span class="ot">&lt;-</span> label_df<span class="sc">$</span>Cell_names</span></code></pre></div>
<p>Next we perform PCA on the data and extract the proportion of variance explained by each component.</p>
<div class="sourceCode" id="cb124"><pre class="sourceCode numberSource r numberLines"><code class="sourceCode r"><span id="cb124-1"><a href="practical-principal-component-analysis.html#cb124-1"></a>sc_pca <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(pollen_mat)</span>
<span id="cb124-2"><a href="practical-principal-component-analysis.html#cb124-2"></a></span>
<span id="cb124-3"><a href="practical-principal-component-analysis.html#cb124-3"></a><span class="co"># variance is the square of the standard deviation</span></span>
<span id="cb124-4"><a href="practical-principal-component-analysis.html#cb124-4"></a>pr_var <span class="ot">&lt;-</span> sc_pca<span class="sc">$</span>sdev<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb124-5"><a href="practical-principal-component-analysis.html#cb124-5"></a></span>
<span id="cb124-6"><a href="practical-principal-component-analysis.html#cb124-6"></a><span class="co"># compute the variance explained by each principal component</span></span>
<span id="cb124-7"><a href="practical-principal-component-analysis.html#cb124-7"></a>prop_var_exp <span class="ot">&lt;-</span> pr_var <span class="sc">/</span> <span class="fu">sum</span>(pr_var)</span></code></pre></div>
<p>Think about the calculation and what exactly it means. We can visualise this</p>
<div class="sourceCode" id="cb125"><pre class="sourceCode numberSource r numberLines"><code class="sourceCode r"><span id="cb125-1"><a href="practical-principal-component-analysis.html#cb125-1"></a>var_exp <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">variance =</span> prop_var_exp, <span class="at">pc =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(prop_var_exp))</span>
<span id="cb125-2"><a href="practical-principal-component-analysis.html#cb125-2"></a></span>
<span id="cb125-3"><a href="practical-principal-component-analysis.html#cb125-3"></a><span class="fu">ggplot</span>(var_exp[<span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>, ], <span class="fu">aes</span>(<span class="at">x =</span> pc, <span class="at">y =</span> variance)) <span class="sc">+</span></span>
<span id="cb125-4"><a href="practical-principal-component-analysis.html#cb125-4"></a>    <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb125-5"><a href="practical-principal-component-analysis.html#cb125-5"></a>    <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Principal Component"</span>,</span>
<span id="cb125-6"><a href="practical-principal-component-analysis.html#cb125-6"></a>         <span class="at">y  =</span> <span class="st">"Variance explained"</span>)</span></code></pre></div>
<div class="inline-figure"><img src="PCA_files/figure-html/pca-scree-1.png" width="95%" style="display: block; margin: auto;"></div>
<p>We see that the first few principal components explain significant variance, but after about the PC10, there is very limited contribution. Next we will plot the data using the first two Principal components as before.</p>
<div class="sourceCode" id="cb126"><pre class="sourceCode numberSource r numberLines"><code class="sourceCode r"><span id="cb126-1"><a href="practical-principal-component-analysis.html#cb126-1"></a>sc_pca_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(sc_pca<span class="sc">$</span>x, <span class="at">cell =</span> <span class="fu">rownames</span>(sc_pca<span class="sc">$</span>x),</span>
<span id="cb126-2"><a href="practical-principal-component-analysis.html#cb126-2"></a>                        <span class="at">var_exp =</span> prop_var_exp)</span>
<span id="cb126-3"><a href="practical-principal-component-analysis.html#cb126-3"></a></span>
<span id="cb126-4"><a href="practical-principal-component-analysis.html#cb126-4"></a><span class="fu">ggplot</span>(sc_pca_df, <span class="fu">aes</span>(<span class="at">x =</span> PC1, <span class="at">y =</span> PC2, <span class="at">col =</span> cell)) <span class="sc">+</span></span>
<span id="cb126-5"><a href="practical-principal-component-analysis.html#cb126-5"></a>    <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb126-6"><a href="practical-principal-component-analysis.html#cb126-6"></a>    <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"bottom"</span>)</span></code></pre></div>
<div class="inline-figure"><img src="PCA_files/figure-html/sc-pca_plot-1.png" width="95%" style="display: block; margin: auto;"></div>
<p>Why is it not useful to create biplot for this example?</p>
<p><em>Hint: try plotting one to see</em></p>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="practical-linear-regression.html"><span class="header-section-number">8</span> Practical: Linear regression</a></div>
<div class="next"><a href="practical-multiple-regression.html"><span class="header-section-number">10</span> Practical: Multiple regression</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#practical-principal-component-analysis"><span class="header-section-number">9</span> Practical: Principal component analysis</a></li>
<li><a class="nav-link" href="#data-1"><span class="header-section-number">9.1</span> Data</a></li>
<li><a class="nav-link" href="#introduction"><span class="header-section-number">9.2</span> Introduction</a></li>
<li><a class="nav-link" href="#example-pca-on-the-mtcars-dataset"><span class="header-section-number">9.3</span> Example: PCA on the mtcars dataset</a></li>
<li><a class="nav-link" href="#creating-plots-for-pca"><span class="header-section-number">9.4</span> Creating plots for PCA</a></li>
<li><a class="nav-link" href="#exercise-i"><span class="header-section-number">9.5</span> Exercise I</a></li>
<li><a class="nav-link" href="#example-single-cell-data"><span class="header-section-number">9.6</span> Example: Single cell data</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/anasrana/ems-practicals/blob/master/PCA.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/anasrana/ems-practicals/edit/master/PCA.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Essentials of Mathematics and Statistics</strong>" was written by Anas A Rana. It was last built on 2024-10-03.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer><!-- dynamically load mathjax for compatibility with self-contained --><script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script><script type="text/x-mathjax-config">const popovers = document.querySelectorAll('a.footnote-ref[data-toggle="popover"]');
for (let popover of popovers) {
  const div = document.createElement('div');
  div.setAttribute('style', 'position: absolute; top: 0, left:0; width:0, height:0, overflow: hidden; visibility: hidden;');
  div.innerHTML = popover.getAttribute('data-content');

  var has_math = div.querySelector("span.math");
  if (has_math) {
    document.body.appendChild(div);
    MathJax.Hub.Queue(["Typeset", MathJax.Hub, div]);
    MathJax.Hub.Queue(function() {
      popover.setAttribute('data-content', div.innerHTML);
      document.body.removeChild(div);
    })
  }
}
</script>
</body>
</html>
